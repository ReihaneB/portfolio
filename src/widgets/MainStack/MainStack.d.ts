interface Stack {
  id: number;
  name: string;
}

export interface MainStackProps {
  stack: Stack[];
}
