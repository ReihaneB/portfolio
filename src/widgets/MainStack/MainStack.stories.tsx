import type { Meta, StoryObj } from '@storybook/react';
import { expect, within } from '@storybook/test';

import { lokiPlayDecorator } from '@/testUtils/loki/play-decorator';

import MainStack from './MainStack';

const meta: Meta<typeof MainStack> = {
  title: 'Widgets/MainStack',
  component: MainStack,
};

export default meta;
type Story = StoryObj<typeof MainStack>;

const defaultArgs = {
  stack: [
    {
      id: 3,
      name: 'Next',
    },
    {
      id: 2,
      name: 'Express',
    },
    {
      id: 1,
      name: 'Postgres',
    },
  ],
};

export const Playground: Story = {
  args: {
    ...defaultArgs,
  },
};

export const _1Element: Story = {
  args: {
    stack: [
      {
        id: 1,
        name: 'Postgres',
      },
    ],
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement }) => {
      const canvas = within(canvasElement);

      expect(args.stack).toHaveLength(1);
      expect(canvas.getAllByTestId('tag-component')).toHaveLength(1);
    }
  ),
};

export const _2Elements: Story = {
  args: {
    stack: [
      {
        id: 1,
        name: 'Postgres',
      },
      {
        id: 2,
        name: 'Express',
      },
    ],
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement }) => {
      const canvas = within(canvasElement);

      expect(args.stack).toHaveLength(2);
      expect(canvas.getAllByTestId('tag-component')).toHaveLength(2);
    }
  ),
};

export const _3Elements: Story = {
  args: {
    stack: [
      {
        id: 1,
        name: 'Postgres',
      },
      {
        id: 2,
        name: 'Express',
      },
      {
        id: 3,
        name: 'Next',
      },
    ],
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement }) => {
      const canvas = within(canvasElement);

      expect(args.stack).toHaveLength(3);
      expect(canvas.getAllByTestId('tag-component')).toHaveLength(3);
    }
  ),
};

export const _4Elements: Story = {
  args: {
    stack: [
      {
        id: 1,
        name: 'Postgres',
      },
      {
        id: 2,
        name: 'Express',
      },
      {
        id: 3,
        name: 'Next',
      },
      {
        id: 4,
        name: 'React',
      },
    ],
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement }) => {
      const canvas = within(canvasElement);

      expect(args.stack).toHaveLength(4);
      expect(canvas.getAllByTestId('tag-component')).toHaveLength(4);
    }
  ),
};

export const MoreThan4Elements: Story = {
  name: 'Maxium 4 elements are shown',
  args: {
    stack: [
      {
        id: 1,
        name: 'Postgres',
      },
      {
        id: 2,
        name: 'Express',
      },
      {
        id: 3,
        name: 'Next',
      },
      {
        id: 4,
        name: 'React',
      },
      {
        id: 5,
        name: 'TypeScript',
      },
    ],
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement }) => {
      const canvas = within(canvasElement);

      expect(args.stack).toHaveLength(5);
      expect(canvas.getAllByTestId('tag-component')).toHaveLength(4);
    }
  ),
};
