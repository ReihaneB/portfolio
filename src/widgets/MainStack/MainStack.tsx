import Tag from '@/components/Tag/Tag';

import type { MainStackProps } from './MainStack.d';
import styles from './MainStack.module.css';

function MainStack({
  stack,
  ...rest
}: MainStackProps) {
  const maxElementsDisplayed = 4;

  return (
    <ul
      className={styles.root}
      {...rest}
    >
      {stack
        .slice(0, maxElementsDisplayed)
        .map(element => (
          <li key={element.name}>
            <Tag>
              {element.name}
            </Tag>
          </li>
        ))}
    </ul>
  );
}

export default MainStack;
