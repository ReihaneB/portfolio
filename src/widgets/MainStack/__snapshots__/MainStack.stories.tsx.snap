// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widgets/MainStack _1Element play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <ul class="src-widgets-MainStack-MainStack-module__root--dscGt">
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Postgres
      </div>
    </li>
  </ul>
</main>
`;

exports[`Widgets/MainStack _2Elements play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <ul class="src-widgets-MainStack-MainStack-module__root--dscGt">
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Postgres
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Express
      </div>
    </li>
  </ul>
</main>
`;

exports[`Widgets/MainStack _3Elements play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <ul class="src-widgets-MainStack-MainStack-module__root--dscGt">
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Postgres
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Express
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Next
      </div>
    </li>
  </ul>
</main>
`;

exports[`Widgets/MainStack _4Elements play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <ul class="src-widgets-MainStack-MainStack-module__root--dscGt">
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Postgres
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Express
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Next
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        React
      </div>
    </li>
  </ul>
</main>
`;

exports[`Widgets/MainStack MoreThan4Elements play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <ul class="src-widgets-MainStack-MainStack-module__root--dscGt">
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Postgres
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Express
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Next
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        React
      </div>
    </li>
  </ul>
</main>
`;

exports[`Widgets/MainStack Playground smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <ul class="src-widgets-MainStack-MainStack-module__root--dscGt">
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Next
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Express
      </div>
    </li>
    <li>
      <div class="src-components-Tag-Tag-module__rootStyle--Cd2YJ"
           data-testid="tag-component"
           tabindex="-1"
      >
        Postgres
      </div>
    </li>
  </ul>
</main>
`;
